conans/requirements.txt

```
@@ -8,7 +8,7 @@ six>=1.10.0
 node-semver==0.6.1
 distro>=1.0.2, <1.2.0
 pylint>=1.9.3,!=2.3.0
-future==0.16.0
+future>=0.16.0
```

I would add an upper limit. It's pretty unlikely that `future` will release a `1.0.0` version given their versions history and with py2 being almost deprecated, but if they did, I'm sure it would be breaking.


```suggestion
future>=0.16.0, <1.0.0
```

---

The same argument could be made for `six`, `fasteners`, and `astroid`. Astroid has already gone from >=1.6.5 to 2.2.5 without causing issues with conan. 

I'm on the fence with this. Yes, a major version change *may* indicate a breaking change, but that's only an assumption as the `astroid` example shows above. 

This package in particular still hasn't had a major release as you point out, so they may simply release version 1.0 to celebrate reaching a milestone.

In general, my feelings on pinning dependencies is that you should only exclude versions that you *know* don't work. It reduces inter-dependency conflicts, and allows security / other important issue fixes to flow through to end user environments without having to release a new version of your package.

With all of that said, I'm also happy to make the change if you really want it 👍

---

I agree with you that the same argument applies to `six`, `fasteners` and `astroid`, and IMO those should have an upper limit too (not for this PR).

The problem I see with Conan is that we cannot fix something in the last version and recommend everyone to upgrade Conan (I wish we could), so sometimes we have to backport this kind of fixes to many releases in the past.

Let's ping @lasote, he will have the final word about this. I also want to know if we have a clear policy on this.

---

By definition of semver, the next breaking will be 0.18. This should be the upper limit.

For the other dependencies, yes, we should probably add some upper limits too, but in a different pull request. Actually I am not very concerned about six and astroid, we will get rid of them in Conan 2.0

---

@memsharded I don't want to be a pedant, but [the definition of semver](https://semver.org/) is that a minor version is incremented:

> when you add functionality in a backwards compatible manner

It looks like 0.18 is largely going to be a bug fix release, and follow the above definition, as 0.17 did: https://github.com/PythonCharmers/python-future/issues?q=label%3A0.18+is%3Aclosed

0.18 could drop anytime soon (https://github.com/PythonCharmers/python-future/issues/478#issuecomment-509410678), which would put us in the exact same position as we were in before I raised this issue & PR - `pefile` will take 0.18 and `conan` won't work for us again 😢 

https://github.com/PythonCharmers/python-future/blob/master/docs/whatsnew.rst

---

The definition of semver (https://semver.org/) excludes versions starting with 0.X.Y to follow that rule:

> Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.

So anything could break :)

Great that you did the checks, thanks very much! And if it seems that 0.18 will be backwards compatible, then I am totally fine including it in the valid range. The thing is that new releases including in the range that are breaking would be breaking Conan and force us to quickly release patches (maybe even for previous Conan versions), while a restrictive range might be inconvenient and create some incompatibilities for some users using other versions, but not for the majority of users. This is why I typically prefer to not increase the ranges too much.

@lasote will take the final decision. Thanks for all the feedback!

---

Include 0.18 in the range.